src/components/lab/NumberInput/NumberInput.tsx

```diff
@@ -0,0 +1,320 @@
+
+    const innerControlRef = React.useRef<HTMLInputElement>(null);
+    const fieldRef = useFormResetHandler({
+        initialValue: defaultValue,
+        onReset: onUpdate ?? voidFunction,
+    });
+    const handleRef = useForkRef(props.controlRef, innerControlRef, fieldRef);
+
+    const handleIncrement = () => {
+        if (canIncrementNumber) {
+            onUpdate?.(
+                clampToNearestStepValue({value: safeValue + step, step: baseStep, min, max}),
+            );
+        }
+    };
+
+    const handleDecrement = () => {
+        if (canDecrementNumber) {
+            onUpdate?.(
+                clampToNearestStepValue({value: safeValue - step, step: baseStep, min, max}),
+            );
+        }
+    };
+
+    const handleWheel: React.WheelEventHandler<HTMLInputElement> = (e) => {
+        const delta = e.shiftKey ? e.deltaX : e.deltaY;
+        if (delta > 0) {
+            handleIncrement();
+        } else if (delta < 0) {
+            handleDecrement();
+        }
+    };
+
+    const handleKeyDown: React.KeyboardEventHandler<HTMLInputElement> = (e) => {
+        if (e.key === KeyCode.SHIFT) {
+            setStep(baseStep * shiftMultiplier);
+        } else if (e.key === KeyCode.ARROW_DOWN) {
+            handleDecrement();
+        } else if (e.key === KeyCode.ARROW_UP) {
+            handleIncrement();
+        }
+        onKeyDown?.(e);
+    };
+
+    const handleKeyUp: React.KeyboardEventHandler<HTMLInputElement> = (e) => {
+        if (e.key === KeyCode.SHIFT) {
+            setStep(baseStep);
+        }
+        onKeyUp?.(e);
+    };
+
+    const handleFocus: React.FocusEventHandler<HTMLInputElement> = (e) => {
+        setActive(true);
+        onFocus?.(e);
+    };
+
+    const handleBlur: React.FocusEventHandler<HTMLInputElement> = (e) => {
+        setActive(false);
+        setStep(baseStep);
+        if (clamp && value) {
+            const clampedValue = clampToNearestStepValue({
+                value,
+                step: baseStep,
+                min,
+                max,
+            });
+            if (value !== clampedValue) {
+                onUpdate?.(clampedValue);
+            }
+        }
+        onBlur?.(e);
+    };
+
+    const handleUpdate = (v: string) => {
+        setInputValue(v);
+        const preparedStringValue = getPossibleNumberSubstring(v, allowDecimal);
+        updateCursorPosition(innerControlRef, v, preparedStringValue);
+        const {valid, value: parsedNumberValue} = getParsedValue(preparedStringValue);
+        if (valid && parsedNumberValue !== value) {
+            onUpdate?.(parsedNumberValue);
+        }
+    };
+
+    const handleInput: React.FormEventHandler<HTMLInputElement> = (e) => {
+        const preparedStringValue = getPossibleNumberSubstring(e.currentTarget.value, allowDecimal);
+        updateCursorPosition(innerControlRef, e.currentTarget.value, preparedStringValue);
```

Comment:
When you increase the value (press the up arrow), the cursor moves to the beginning and then to the end.